File: 代码随想录/数组/27移除元素.py
# 给你一个数组 nums 和一个值 val，你需要 原地 移除所有数值等于 val 的元素。元素的顺序可能发生改变。然后返回 nums 中与 val 不同的元素的数量。
from typing import List
nums = [0, 1, 2, 2, 3, 0, 4, 2]
val =  2
def removeElement_corrected_brute_force(nums, val) -> int:
    i= 0
    current_size = len(nums)
    while i<current_size:
        if nums[i] == val:
            for j in range(i+1,current_size):
                nums[j-1] = nums[j]
            current_size -=1
            # 注意：此时 i 不需要增加，因为要检查新移动到 i 位置的元素
        else:
            i+=1
    return current_size
    
def removeElement(nums: List[int], val: int) -> int:
        while val in nums: # 1. 检查 val 是否还在 nums 中，这一步遍历nums O(n)
            nums.remove(val) # 2. 如果在，移除第一个找到的 val ，这一步遍历nums O(n) 实际上O(N²) 
        return len(nums)
    
def shuangremoveElement(nums: List[int], val: int) -> int:
    slow = 0
    for fast in range(len(nums)):
        if nums[fast]!=val:
            nums[slow]=nums[fast]
            slow+=1
    return slow
print(removeElement_corrected_brute_force(nums=nums,val=val))

print(shuangremoveElement(nums=nums,val=val))

File: 代码随想录/数组/27移除元素.cpp
#include <cstddef>  // For size_t, which is the type returned by sizeof
#include <iostream> // For std::cout if you want to print
#include <vector>

int removeElement(std::vector<int> &nums, int val)
{
    int slow = 0;
    for (int fast = 0; fast < nums.size(); fast++)
    {
        if (nums[fast] != val)
        {
            nums[slow++] = nums[fast];
        }
    }
    return slow;
}
int removeElement2(std::vector<int> &nums, int val)
{
    int n = nums.size();
    int left = 0;
    for (int right = 0; right < n; right++)
    {
        if (nums[right] != val)
        {
            nums[left] = nums[right];
            left++;
        }
    }
    return left;
}
int removeElement3(std::vector<int> &nums, int val)
{
    // 当左指针 left 指向的元素是我们想要移除的 val 时，我们尝试用右指针 right 指向的元素去覆盖它。
    // 这样做的目的是尽量把“有效”的元素往前挪，把“无效”的元素（或者暂时不关心的元素）往后挪。
    int left = 0, right = nums.size();
    while (left < right)
    {
        if (nums[left] == val)
        {
            nums[left] = nums[right - 1];
            right--;
        }
        else
        {
            left++;
        }
    }
    return left;
}

int main()
{
    int nums[8]{0, 1, 2, 2, 3, 0, 4, 2};
    int val = 2;

    size_t array_size = sizeof(nums) / sizeof(nums[0]);

    int i = 0;

    int current_size = array_size; // Corrected line and added semicolon

    while (i < current_size)
    {
        /* code */
        if (nums[i] == val)
        {
            /* code */
            for (int j = i + 1; j < current_size; j++)
            {
                nums[j - 1] = nums[j];
            }
            current_size--;
        }
        else
        {
            i++;
        }
    }
    std::cout << current_size << std::endl;
    std::vector<int> nums1 = {0, 1, 2, 2, 3, 0, 4, 2};
    std::vector<int> nums2 = {0, 1, 2, 2, 3, 0, 4, 2};
    std::vector<int> nums3 = {0, 1, 2, 2, 3, 0, 4, 2};
    std::cout << removeElement(nums1, 2) << std::endl;
    std::cout << removeElement2(nums2, 2) << std::endl;
    std::cout << removeElement2(nums3, 2) << std::endl;
}
